feat: allow using custom R1CS to QAP reductions

gakonst opened the pull request:
## Description

As discussed, it is potentially useful to allow customizing the R1CS to QAP conversion, for cases such as Circom's where they skip some of the FFTs.

## Implementation

This PR abstracts the R1CS to QAP to a trait and exposes methods for providing implementors of the trait. It maintains backwards compat.

Pratyush commented on src/r1cs_to_qap.rs:
```diff
@@ -41,12 +41,23 @@ where
 
-pub(crate) struct R1CStoQAP;
+pub struct R1CStoQAP;
 
-impl R1CStoQAP {
+pub trait QAPCalculator {
```
Maybe I would rename the trait to `R1CSToQAP`, and change the struct name to something more specific; maybe `LibsnarkReduction`, and the new circom one would be `CircomReduction`?

Pratyush reviewed (changes requested):
LGTM modulo the naming change!

Pratyush commented:
Actually another suggestion: maybe we can parameterize the Groth16 struct itself by a new generic type parameter bounded by the `R1CSToQap` trait, and set the default to the old version? Do you think that would be cleaner?

gakonst commented:
Addressed your comments!

> Actually another suggestion: maybe we can parameterize the Groth16 struct itself by a new generic type parameter bounded by the `R1CSToQap` trait, and set the default to the old version? Do you think that would be cleaner?

I think we should be considering a larger refactor (which I've started [here](https://github.com/gakonst/groth16/compare/feat/customizable-r1cs-to-qap...gakonst:feat/cache-intermediate-steps?expand=1)) so as to allow the prover to cache intermediate steps, which should also help with having a cleaner abstraction around the QAP type

Pratyush commented on src/r1cs_to_qap.rs:
```diff
@@ -10,7 +10,7 @@ use core::ops::{AddAssign, Deref};
 
 #[inline]
-fn evaluate_constraint<'a, LHS, RHS, R>(terms: &'a [(LHS, usize)], assignment: &'a [RHS]) -> R
+pub fn evaluate_constraint<'a, LHS, RHS, R>(terms: &'a [(LHS, usize)], assignment: &'a [RHS]) -> R
```
```suggestion
/// Computes the inner product of `terms` with `assignment`.
pub fn evaluate_constraint<'a, LHS, RHS, R>(terms: &'a [(LHS, usize)], assignment: &'a [RHS]) -> R
```

Pratyush commented on src/r1cs_to_qap.rs:
```diff
@@ -41,12 +41,30 @@ where
 }
 
-pub(crate) struct R1CStoQAP;
+pub trait R1CStoQAP {
```
```suggestion
/// Computes instance and witness reductions from R1CS to 
/// Quadratic Arithmetic Programs (QAPs).
pub trait R1CStoQAP {
```

Pratyush commented on src/r1cs_to_qap.rs:
```diff
@@ -41,12 +41,30 @@ where
 
-pub(crate) struct R1CStoQAP;
+pub trait R1CStoQAP {
+    fn instance_map_with_evaluation<F: PrimeField, D: EvaluationDomain<F>>(
```
```suggestion
	/// Computes a QAP instance corresonding to the R1CS instance defined by `cs`.
    fn instance_map_with_evaluation<F: PrimeField, D: EvaluationDomain<F>>(
```

Pratyush commented on src/r1cs_to_qap.rs:
```diff
@@ -41,12 +41,30 @@ where
+        t: &F,
+    ) -> Result<(Vec<F>, Vec<F>, Vec<F>, F, usize, usize), SynthesisError>;
+
+    fn witness_map<F: PrimeField, D: EvaluationDomain<F>>(
```
```suggestion
	/// Computes a QAP witness corresponding to the R1CS witness defined by `cs`.
    fn witness_map<F: PrimeField, D: EvaluationDomain<F>>(
```

Pratyush commented on src/r1cs_to_qap.rs:
```diff
@@ -41,12 +41,30 @@ where
+    ) -> Result<Vec<F>, SynthesisError>;
+}
+
+pub(crate) struct LibsnarkReduction;
```
```suggestion
/// Computes the R1CS-to-QAP reduction defined in [`libsnark`](https://github.com/scipr-lab/libsnark/blob/2af440246fa2c3d0b1b0a425fb6abd8cc8b9c54d/libsnark/reductions/r1cs_to_qap/r1cs_to_qap.tcc).
pub struct LibsnarkReduction;
```

Pratyush commented on src/r1cs_to_qap.rs:
```diff
@@ -41,12 +41,30 @@ where
+        prover: ConstraintSystemRef<F>,
+    ) -> Result<Vec<F>, SynthesisError>;
+
+    fn h_query_scalars<F: PrimeField, D: EvaluationDomain<F>>(
```
Not sure what the appropriate documentation for this should be. cc @kobigurk?

Pratyush commented:
Also, for the documentation, should we add the details of the reduction as described in libsnark: https://github.com/scipr-lab/libsnark/blob/2af440246fa2c3d0b1b0a425fb6abd8cc8b9c54d/libsnark/reductions/r1cs_to_qap/r1cs_to_qap.tcc

Pratyush commented:
I updated the API names slightly @gakonst, but this is good to merge otherwise. Any comments?
